### PR TITLE
Mark containerRuntime as dirty while uploading blob

### PR DESCRIPTION
**Pull request by dannimad**
## Description

It seems like a pretty common concern for hosts to want to know if a fluid document has blobs pending upload, so that they can present some UI informing the user on attempted close.

After discussions with @Abe27342 and @anthony-murphy we believe marking the containerRuntima while uploading blobs is the best solution for hosts to determine whether or not there will be data loss if error or disconnections happens.

## Reviewer Guidance

> The change is mainly making sure container is dirty while blob uploading plus some caveats to avoid asserts to be triggered and adding a unit tets for dirty flag checks while blobs uploading process.

### CODE REVIEW

**Comment by agarwal-navin** [comment]
Can you please add a description. Adding a link to ADO item isn't enough because folks outside MS can't see it. Also, a description helps reviewer review the changes.

**Comment by agarwal-navin** [comment]
Looks like you created your branch in the microsoft/FluidFramework repo. The recommended way is to create branches in your private fork.

**Comment by dannimad** [comment]
Description added. And I've been creating branches from microsoft/FluidFramework for a while. Probably a bad practice. I will keep this in mind and create branches from my fork for now on.

**Comment by agarwal-navin** [comment]
> Description added. And I've been creating branches from microsoft/FluidFramework for a while. Probably a bad practice. I will keep this in mind and create branches from my fork for now on.

Thank you! Please note that you should still add link to ADO items as AB#<ado_item_number> if there is one. This will create a link for our internal usage from the PR <-> ADO item.

Also, see https://github.com/microsoft/FluidFramework/wiki/Repo-Basics for details on how to setup your environment.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -2632,7 +2632,12 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
 
     public async uploadBlob(blob: ArrayBufferLike): Promise<IFluidHandle<ArrayBufferLike>> {
         this.verifyNotClosed();
-        return this.blobManager.createBlob(blob);
+        const previousDirtyState = this.isDirty;
+        this.emitDirtyDocumentEvent = true;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2636]: I do not understand what this code does or tries to do. I want to point out that there could be many calls to this API in parallel, so any modification of global state that does not accumulate (i.e. counters vs. boolean) is very likely a bug. Plus changes in state that are not revered at the end of the process (this.emitDirtyDocumentEvent is changed to true and never reset - not sure what's the thinking here)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -2590,12 +2590,12 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
     }
 
-    private updateDocumentDirtyState(dirty: boolean) {
+    private updateDocumentDirtyState(dirty: boolean, uploadingBlob: boolean = false) {
         if (this.attachState !== AttachState.Attached) {
             assert(dirty, 0x3d2 /* Non-attached container is dirty */);
         } else {
             // Other way is not true = see this.isContainerMessageDirtyable()
-            assert(!dirty || this.hasPendingMessages(),
+            assert(!dirty || this.hasPendingMessages() || uploadingBlob,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2598]: I want to understand better thinking here and scoping.

I'm not sure we need that (design).
I'd go with a solution where dirty flag is fully controlled by ops.
Blob upload generates op and operation is not completed until op is acked.

You can argue that we mark document dirty when operation starts, but this is really undefined - how does one define "start"? There is no requirement anywhere in the system that when first part (synchronous part of the call) is over, it's a "start". And given that (at least current) design does not apply this change immediately to date mode, data model is actually never dirty through the whole process. It's atomically switched to "image is there" when op is acked.

Maybe I do not fully understand concerns, but they feel like similar to "I have something in the app happening in response to user action, data model is not yet dirty, but I want it dirty". It does feel like it's not really runtime problem.

Maybe it's only for offline mode? If so, I really want to call out that it's not ideal to have two different models RE how changes are applied to data model (synchronous & async). I'd want to have a discussion - can we have same flow and not complicate designs for one side of implementation?

If we had fully synchronous flow for image to apply to local data mode, then we would likely need something like that, but (as comment below points out) you need to work with counters, not booleans, similar to how this.hasPendingMessages() works - it looks a ALL pending ops. All of this code dealing with blobs should look at ALL pending blob uploads.

File: packages/test/test-end-to-end-tests/src/test/blobs.spec.ts [hunk 3]
```diff
@@ -129,6 +129,29 @@ describeFullCompat("blobs", (getTestObjectProvider) => {
...
+        assert.equal(container1.isDirty, true);
+
+        const container2 = await provider.loadTestContainer(testContainerConfig);
+        const dataStore2 = await requestFluidObject<ITestDataObject>(container2, "default");
+
+        await provider.ensureSynchronized(this.timeout() / 3);
+        assert.equal(container1.isDirty, false);
+        const blobHandle = dataStore2._root.get<IFluidHandle<ArrayBufferLike>>(testKey);
+        assert(blobHandle);
+        assert.strictEqual(bufferToString(await blobHandle.get(), "utf-8"), testString);
```
Comment (vladsud) [inline packages/test/test-end-to-end-tests/src/test/blobs.spec.ts:152]: Please ensure there is coverage for 2-3 parallel blob uploads and that system maintains correct state through these transitions and gets back to normal state after full flow.
I think you need to ensure that for online, offline and a mix of offline & online modes (i.e. losing connection halfway through second-third upload).
This is clearly should have covered even before your changes (so maybe we should ask Wes for help), but I worry these changes are going to create issues in those cases (see other comments)

**Review by vladsud — CHANGES_REQUESTED** [review]
Please see my comments - I want to better understand intention (do we need it?), if it's mostly about subset of feature (offline mode), and ensure that concurrency is properly handled

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (dannimad) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2636]: For some reason I believed the emitDirtyDocumentEvent was changed during updateDocumentDirtyState but I see it does not. That could be easily fixed by preserving the previous state. I agree asynchoronous calls are a problem in this implementation. What I could do is to check BlobManager.pendingBlobs size in line 1111 (constructor) but not sure we will be able to immediately flag the container as dirty like I'm doing it in this implementation.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (dannimad) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2598]: Does changing solution to check for BlobManager.pendingBlobs array solve your concerns about sync vs async flow? I'm still not completely sure that we should only check isDirty state during containerRuntime constructor though

File: packages/test/test-end-to-end-tests/src/test/blobs.spec.ts (hunk 3 above)
Comment (dannimad) [reply packages/test/test-end-to-end-tests/src/test/blobs.spec.ts:152]: Agree. Parallel blob uploading is not covered by any unit tests afaik. Once I can define an async solution I'll add corresponding unit tests

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 4]
```diff
@@ -287,6 +288,7 @@ export class BlobManager {
     }
 
     public async createBlob(blob: ArrayBufferLike): Promise<IFluidHandle<ArrayBufferLike>> {
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/blobManager.ts:290]: we should add the concept of dirty to the blob manager itself, and then hook into that from the container runtime, not try to have the runtime independently manage dirty state around blobs. this should also be more unit testable

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2598]: The host needs to know about inflight blob operations, otherwise it can lead to dataloss. The mechanism we have for the host to determine if there are inflight operation on a document is save/dirty

**Comment by anthony-murphy** [comment]
> Description added. And I've been creating branches from microsoft/FluidFramework for a while. Probably a bad practice. I will keep this in mind and create branches from my fork for now on.

you should close this PR and restart to a branch in your fork. we can't continue creating branches in the main repo.

**Review by anthony-murphy — CHANGES_REQUESTED** [review]
re-create on a branch in your fork, and delete this branch in the main repo

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 4 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/blobManager.ts:290]: I'd disagree. Dirty means - there are changes that were already applied and reflected in data model (and thus very likely - on the screen) that were not saved. Pending blob upload does not satisfy that criteria - there are no changes to local data model until upload fully commits to data model.
Exception is offline blob creation flow.

Comment (vladsud) [reply packages/runtime/container-runtime/src/blobManager.ts:290]: Litmus test for me - are changes on the screen only for local client? If they are not, or if they are reflected on remote machine, then document is not dirty. If app has a layer of "lies" on top of local FF state, then they need to incorporate such lies into their calculation of dirty state tracking - it's the right layers to have and track such things.

Comment (dannimad) [reply packages/runtime/container-runtime/src/blobManager.ts:290]: I guess the issue then is the definition of "dirty". We had discussed this, and we got that "a state when there is potential data loss if container is closed" could be a good one. So having the runtime as dirty while uploadBlob made sense. We discussed also having a separate runtime API like runtime.hasPendingBlobs but we decided dirty option was more convenient (single property for user to check for potential data loss state)

Comment (vladsud) [reply packages/runtime/container-runtime/src/blobManager.ts:290]: If we want to push into this direction, I'd love to see the following things happened:
Saving container state would preserve these blobs (or rehydration) in all cases. In other words, when we communicate dirty, saved document state has local changes (not empty). Including cases like
- blobs uploaded in "attaching" state to be included in getPendingBlobs()
- If blobs pending upload (including above case) are saved in getPendingBlobs(), they are applied to data model on container rehydration. It's not the case today, because uploadBlob() is async API (and caller has no result to hold on to while upload is pending), so it does not matter - if we consider container is dirty in this state, then we acknowledge that we have data loss.

I have a feeling that fixing cases above forces us toward a model that follows "transition to offline" mode (from BlobManager perspective) when saving container. And if it's the case, maybe it's better to switch to fully sync model (i.e. blob uploads are synchronous from model perspective, upload block all the changes that follows).

The assumption I have here - apps (like Outlook, WB) who use dirty event to save local state have consistent view into the world.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2598]: @dannimad - yes, that would be the better place, but I have a lot concerns here (see below long comment with a bunch of replies). Moving it to blob manage is just first step. We should re-evaluate all transition and make sure that system has some consistency. This includes small things (like treating blobs uploaded in "attaching" and "attached" states the same), but also big things (if it's dirty, then set of local changes is non-empty).

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 4 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/blobManager.ts:290]: @dannimad, please define data loss. If I save local state of container (Container.closeAndGetPendingLocalState) and do not have these changes, is this data loss? If not (this is today's behavior) - than container should not be dirty. If it's data loss (and we indicate it so through dirty flag), then I expect data to be saved and rehydrated properly - that's what our clients would expect. But that's not possible in today's system without changing all uploads to apply to local model synchronously - either immediately, or by transitioning to "offline" mode on getPendingBlobs(). But latter does not help, as no changes yet were applied to local data, so it's still a data loss. So, we have to always operate in sync mode.

Basically, you can't change just one piece of the puzzle - the whole castle starts to collapse :) I really want to see consistent model here. It's already not very consistent (due to transitionToOffline()), but I have a hard time seeing why we need to make it even less consistent and hard to understand (mostly - from how it's projected to the users).

I think you look at it only through perspective of online container, with an app that has a layer of "lies" on top of it (like Bohemia - they have a layer that applies image to presentation layer), and container that never gets saved. I want to see at full picture (all workflows) and ensure we present consistent world through our APIs.
